### PR TITLE
New search standard and user search

### LINKED ISSUE #2657: Provide JS-Only version of OCA.Search
It would be good to have a js-only version of OCA.Search.
OCA.Search requires you to implement PHP APIs at the moment.
This is not practical for apps like Calendar or Contacts, which don't store the data on themselves.

https://github.com/nextcloud/server/blob/b4f36d44c43aac0efdc6c70ff8e46473341a9bfe/lib/private/Search.php#L59L82
https://github.com/nextcloud/server/blob/d2fab66d49b6f13b9afd4817f064d8531303142f/core/search/js/search.js#L135L142

### PR DESCRIPTION
**Pull request by skjnldsv**
This is a tricky one!
Fixes #2657 

- Added support for user search in the new users vue list
- Added support for simple registration of a search in javascript
  - `new OCA.Search(searchfunction, resetfunction)`
  - Any app can register this with plain pure js and simply add its manager.
- Deprecated `OCA.Search` is now `OCA.Search.Core` to ensure devs migrate their search system to this as well.
- Fixed design of the global search in Files

### CODE REVIEW

**Comment by MorrisJobke** [comment]
Conflicts :/

**Review by juliusknorr — APPROVED** [review]
:+1: Works nicely also with multiple instances on one page.

**Comment by MorrisJobke** [comment]
I added one user "test1" and searched for "test", then I got one result, then two, then three and it stopped after 4:

bildschirmfoto 2018-06-19 um 16 15 22
bildschirmfoto 2018-06-19 um 16 15 24

Same happens with two users in the result set:

bildschirmfoto 2018-06-19 um 16 16 49

**Comment by skjnldsv** [comment]
@MorrisJobke done

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
I found another small issue:

* open the apps management
* type in the search term
* press enter
* before: it did nothing and just searched
* after: it does a POST request resulting in a white page

**Comment by skjnldsv** [comment]
@MorrisJobke the app management need to be updated to this standard as well, I'll handle it in a sec :)

File: core/search/js/search.js [hunk 1]
```diff
@@ -1,421 +1,139 @@
...
-							self.search(query);
-						} else {
-							self.hideResults();
-						}
-						if(self.hasFilter(getCurrentApp())) {
-							self.getFilter(getCurrentApp())(query);
-						}
-					}
+			// Show search
+			document.getElementById('searchbox').style.display = 'block';
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply core/search/js/search.js]: It's a deeper issue as there is no real reason why the search needs to me initiated on the login page :p

**Comment by nickvergessen** [comment]
Since this patch the search icon/bar is visible on any app again, although there might not be search functionality available, eg activity, talk, settings, …

Can we do something about this?

**Comment by skjnldsv** [comment]
@nickvergessen this shouldn't. Who is registering the search on those apps?

**Comment by skjnldsv** [comment]
This registration shouldn't be there on every page :/

**Pull request merged**